compiler/rustc_mir_dataflow/src/impls/live_borrows.rs

```diff
@@ -0,0 +1,1469 @@
+//! cases.
+//!
+//! 1. Dependency Graph
+//!
+//! The `Node`s in the dependency graph include data values of type `NodeKind`. `NodeKind` has
+//! three variants: `Local`, `Borrow` and `LocalWithRefs`.
+//!     ```text
+//!     * `NodeKind::Local` is used for `Local`s that are borrowed somewhere (`_4` in our example), but aren't
+//!        themselves references or pointers.
+//!     * `NodeKind::Borrow` is used for `Local`s that correspond to borrows (`_5` in our example). We equate
+//!        re-borrows with the `Node` that corresponds to the original borrow.
+//!     * `NodeKind::LocalWithRefs` is used for `Local`s that aren't themselves refs/ptrs, but contain
+//!       `Local`s that correspond to refs/ptrs or other `Local`s with `Node`s of kind `NodeKind::LocalWithRef`s.
+//!       `LocalWithRefs` is also used for exposed pointers.
+//!     ```
+//!
+//! Let's look at an example:
+//!
+//! ```text
+//!     _4 = Bar {}
+//!     _5 = Ref(_4)
+//!     _6 = Foo(..)(move _5)
+//!     ...
+//!     _7 = (_6.0)
+//!
+//!     In this example `_6` would be given `NodeKind::LocalWithRefs` and our graph would look
+//!     as follows:
+//!
+//!     `_7` (NodeKind::Borrow) <-> `_6` (NodeKind::LocalWithRefs) <-> `_5` (NodeKind::Borrow) <-> `_4` (NodeKind::Local)
+//!
+//!     On the one hand we need to treat `Local`s with `Node`s of kind `NodeKind::LocalWithRefs` similarly
+//!     to how we treat `Local`s with `Node`s of kind `NodeKind::Local`, in the sense that if they are
+//!     borrowed we want to keep them live over the live range of the borrow. But on the other hand we
+//!     want to also treat them like `Local`s with `Node`s of kind `NodeKind::Borrow` as they ultimately
+//!     could also contain references or pointers that refer to other `Local`s. So we want a
+//!     path in the graph from a `NodeKind::LocalWithRef`s node to the `NodeKind::Local` nodes, whose borrows
+//!     they might contain.
+//!
+//!     Additionally `NodeKind::LocalWithRefs` is also used for raw pointers that are cast to
+//!     `usize`:
+//!
+//!     _4 = Bar {}
+//!     _5 = AddressOf(_4)
+//!     _6 = _5 as usize
+//!     _7 = Aggregate(..) (move _6)
+//!     _8 = (_7.0)
+//!
+//!     In this example our graph would have the following edges:
+//!         * `_5` (Borrow) <-> `_4` (Local)
+//!         * `_6` (LocalWithRefs) <-> `_5` (Borrow)
+//!         * `_7` (LocalWithRefs) <-> `_6` (LocalWithRefs)
+//!         * `_8` (LocalWithRefs) <-> `_7` (LocalWithRefs)
+//!
+//!     We also have to be careful about dealing with `Terminator`s. Whenever we pass references,
+//!     pointers or `Local`s with `NodeKind::LocalWithRefs` to a `TerminatorKind::Call` or
+//!     `TerminatorKind::Yield`, the destination `Place` or resume place, resp., might contain
+//!     these references, pointers or `LocalWithRefs`, hence we have to be conservative
+//!     and keep the `destination` `Local` and `resume_arg` `Local` live.
+//! ```
```

Comment:
I don't understand what this graph models.
In itself, what does an edge mean?
I'd have expected `_5 -> _4` to mean `_5's value contains a pointer to _4`, but that does not fit with your description.